Add display tools for quantities.

Aehmlo opened the pull request:
I'm not completely satisfied with this solution, and I welcome feedback, but it works.

Resolves #13.

iliekturtles commented:
Thanks! First thing I notice is that the new `fmt` mod should be (at least partially) generated by the `system!` macro. Right now formatting only works for the `si` system when it should work for any system. Once this is done tests can be moved into the top-level `tests` module. I'll review in more detail this weekend.

iliekturtles commented on src/si/fmt.rs:
```diff
@@ -0,0 +1,133 @@
+#[cfg(test)]
+mod tests {
+
+    use si::fmt::{Arguments, DisplayStyle::*};
```
`uom` still supports `rustc` back to 1.20.0 which doesn't support nested imports.

iliekturtles commented on src/si/fmt.rs:
```diff
@@ -0,0 +1,133 @@
+        {
+            fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                let value = self.quantity.get::<T>(self.arguments.subunit);
+                write!(f, $pat, value)?;
```
Instead of passing `$pat` in via macro parameter call the value's `fmt::$name` implementation: `value.fmt(f)` (or, if explicit call syntax is needed: `<V as ::lib::fmt::$name>::fmt(v, f)`). See the `Debug` impl for `Quantity` for an example.

iliekturtles commented on src/si/fmt.rs:
```diff
@@ -0,0 +1,133 @@
+    quantity: Quantity<D, U, V>,
+}
+
+trait Convert<V> { // Use this get<N> instead of the (private) get<N> implemented per-unit.
```
Would like to remove this trait. Calling the `from_base` directly should work.

iliekturtles commented on src/system.rs:
````diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+            /// # Usage
+            /// ```
+            /// # use uom::si::Unit;
+            /// # use uom::si::{f32::*, length::{centimeter, meter}, fmt::{Arguments, DisplayStyle::*}};
````
Remove nested imports. I'd like to get tests running for `rustc` 1.20.0 again.

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+            }
+
+            impl<T: Unit> Arguments<T> {
+
```
Remove blank line.

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+                        quantity,
+                    }
+                }
+
```
Remove blank line.

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+            display_arguments!(Octal);
+            display_arguments!(UpperExp);
+            display_arguments!(UpperHex);
+
```
Remove blank line.

iliekturtles commented on src/system.rs:
````diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+            /// assert_eq!("100 centimeters", format!("{}", a.with(length)));
+            /// ```
+            #[allow(missing_debug_implementations)] // Prevent accidental direct use.
+            #[derive(Clone)]
````
Why not `Copy`?

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+                }
+
+                /// Specifies a quantity to display.
+                pub fn with<D, U, V>(&self, quantity: Quantity<D, U, V>) -> QuantityArguments<T, D, U, V>
```
If `Arguments` was `Copy` I think this would be changed to taking self by value and not calling `clone()` down below.

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+                        T: Unit + Conversion<V, T = V::T>,
+                        D: Dimension + ?Sized,
+                        U: Units<V> + ?Sized,
+                        V: Num + Conversion<V> + fmt::$style + Clone,
```
Is `Clone` necessary?

iliekturtles commented on src/system.rs:
```diff
@@ -1071,6 +1071,101 @@ macro_rules! system {
+                    {
+                        fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+                            let value = from_base::<D, U, V, T>(&self.quantity.value);
+                            value.fmt(f).and_then(|_| {
```
I think I liked the original code that used `?` better, e.g.

```rust
value.fmt(f)?;
write!(f, " {}", ...)
```

Aehmlo commented:
Well, it wasn't "the next few days" as I was hoping, but here we are.

I've fixed the above issues and added a mechanism for using `Arguments` directly and indirectly. There are two traits, `Direct` and `Indirect`, that the user can bring into scope in order to choose the behavior they want to use. How does this solution look to you?

iliekturtles commented on src/system.rs:
````diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+            /// ```
+            #[allow(missing_debug_implementations)] // Prevent accidental direct use.
+            #[derive(Clone, Copy)]
+            pub struct Arguments<T: Unit> {
````
Move generics bound for `T` into a where clause.

iliekturtles commented on src/system.rs:
```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+
+            /// Allows for direct and immediate formatting of quantities in the specified style
+            /// and unit.
+            pub trait Direct<T>
```
I'm thinking that using traits to overload `new()` adds too much complexity and just having two methods: `new(...) -> Arguments` and `new_with(...) -> QuantityArguments` makes things easier to use. With traits the trait needs to be in scope  and if you happen to have both in scope you need to use fully qualified method call syntax.

iliekturtles commented on src/system.rs:
```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+                    U: Units<V> + ?Sized,
+                    V: Num + Conversion<V>,
+                    {
+                        let a: Arguments<T> = <Self as Indirect<T>>::new(subunit, style);
```
Directly constructor the result? `QuantityArguments { arguments: Arguments { ... }, quantity }`

Aehmlo replied:
I *believe* the compiler should inline this anyway, and it makes me uneasy to copy-paste code where I could just have a zero-cost call into another entry point. It's unlikely that this struct will need to change, but it's not impossible.

iliekturtles commented on src/system.rs:
```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+                }
+            }
+
+            impl<T: Unit> Arguments<T> {
```
Move generics bound for `T` into a where clause.

iliekturtles commented on src/system.rs:
```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+
+            #[doc(hidden)]
+            #[allow(dead_code)] // We don't use this internally.
+            pub struct QuantityArguments<T: Unit, D: Dimension + ?Sized, U: Units<V> + ?Sized, V: Num + Conversion<V>> {
```
Move generics bounds into a where clause.

iliekturtles commented on src/system.rs:
```diff
@@ -1197,6 +1197,162 @@ macro_rules! system {
+                            value.fmt(f)?;
+                            write!(f, " {}", match self.arguments.style {
+                                DisplayStyle::Abbreviation => T::abbreviation(),
+                                DisplayStyle::Description => if value.is_one() {T::singular()} else {T::plural()},
```
Put spaces on each side of braces: `... else { T::plural() },`

Aehmlo commented:
I've refactored this to simply use `new` and `new_with`; I agree that letting a little bit of the backing  hierarchy leak is better than requiring the user to manage trait import. A couple of lines break the 100-column rule, but I've left them for now, since I don't think they'd really benefit from reworking. I also haven't added tests outside of the doc tests; if you feel that more are needed, I'm happy to do so.

Aehmlo commented:
It looks to me that the build error is #52 again (although it's somewhat strange to see that it only happens on 1.20.0).

iliekturtles commented:
I think performance improvements since 1.20.0 allow newer versions of rustc to complete in less than 10 minutes.

iliekturtles commented:
Turns out we weren't as close as I thought at first. After reviewing in more depth I found a number of needed changes. I started making changes in my `dev-display` branch before I realized the full scope. I suggest squashing my commit into yours and then making the rest of the changes.

- Remove all the `#[allow(dead_code)]` attributes and add tests to `tests.rs` that go into more depth than the doc tests do. Check `Abbreviation` and `Description`. Check that other [format flags](https://doc.rust-lang.org/stable/std/fmt/index.html#syntax) (fill, align, sign, ...) work when used. Add a test for each trait to ensure they're being implemented correctly and pass through to the underlying type's implementation.
- I moved `QuantityArguments` below `Arguments`.
- Doc tests need `cfg_attr`s to only run if `f32` and `si` features are both defined.
- Add a `compile_fail` test to show/ensure that `Arguments` doesn't impl `Debug` (do for `Display` instead?).
- To be consistent with other code I renamed the `T` generic parameter to `N`.
- I renamed "subunit" to "unit."
- Any public methods that have a parameter of type `N` need to be moved into the `quantity!` macro. Currently you can do something like `Arguments::new_with(length, seconds, Description)` and use a `Quantity` and unit that don't match.
  - This also brings up an issue about how to print values for quantities that are not explicitly defined. I'm inclined to punt on this one for the moment.
  - Add a `compile_fail` test to show that mixing quantities and units wont work.

iliekturtles commented:
I don't think so, but can you elaborate some more? In regards to moving code into the `quantity!` macro it's because of the need to use a type's specific `Unit` trait. e.g. you shouldn't be able to use `seconds` as the unit for a quantity of `length`.

Aehmlo commented:
You're right, we can't constrain using `Dimension` as-is because `Unit`s don't contain any information about dimensions. I assume this is intentional, but it seems to me that if we were to use something like an associated type to embed dimensional information into the unit, we would have a nice solution here. Then we could do something like `where N: Unit<Dimension = D>` and eliminate the aforementioned issues.

It's very likely that there's some subtlety I'm missing, or that I'm just not thinking hard enough (it's pretty early for this), but this does seem to be a good solution to me that has some nice unit-safety benefits.

I investigated this glancingly, but I saw some things that might be problematic with `Sized` bounds. This may very well be impossible with the current architecture.

Aehmlo commented:
I'm going to go ahead and close this until I'm able to pick it back up again.

iliekturtles commented:
Just pushed the [work](https://github.com/iliekturtles/uom/tree/dev-display) I've put in so far. Did you want to take these changes and re-open this PR or review what I've done and I can complete the changes?

 * [ ] Review generic parameters.
 * [ ] `pub(crate)` necessary?
 * [ ] Finish comments.
   * [ ] Move examples to `format_args` and `to_format_args`?
 * [ ] Correct tests
   * [ ] `Clone`/`Copy` not derived correctly. Need custom implementations that ignore `Dimension`.
   * [ ] Assert tests for `DisplayStyle`?
   * [ ] Add tests for each trait (`Binary`, `Debug`, `Display`, ...) (Single test that runs through all the traits at once or separate tests?)
   * [ ] Add round trip test `Display` -> `FromStr`.
